File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/SmbConnection.kt
package app.storyarc.feature.library

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.storyarc.core.model.Source
import app.storyarc.core.model.SourceConnectionState
import app.storyarc.core.model.SourceKind
import app.storyarc.core.persistence.CredentialStore
import app.storyarc.core.smb.SmbAddress
import app.storyarc.core.smb.SmbClient
import app.storyarc.core.smb.SmbEntry
import app.storyarc.core.smb.SmbError
import app.storyarc.core.smb.SmbIdentity
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Adding a share, from a host and a password to a saved source.
 *
 * `network-share` validates "before saving" and reports "the specific failure", then lets a
 * reader "browse the share's directory tree and pick the folder to use as the library root".
 * Those are two steps rather than one screen, and this is the state machine between them.
 */
class SmbConnection(
    private val context: Context,
    private val credentials: CredentialStore?,
) : ViewModel() {

    sealed interface Step {
        data object Entering : Step

        data object Connecting : Step

        /** Connected. The reader is now choosing which folder to read. */
        data class Browsing(
            val identity: SmbIdentity,
            val path: String,
            val entries: List<SmbEntry>,
        ) : Step

        data class Failed(val message: String) : Step
    }

    private val _step = MutableStateFlow<Step>(Step.Entering)
    val step: StateFlow<Step> = _step.asStateFlow()

    val host = MutableStateFlow("")
    val share = MutableStateFlow("")
    val username = MutableStateFlow("")
    val password = MutableStateFlow("")

    private var resolved: SmbAddress? = null

    /** Whether there is enough typed in to try. */
    fun isReady(): Boolean = host.value.isNotBlank() && share.value.isNotBlank()

    /**
     * Connects and lists the share's root, which is the first thing a reader has to choose
     * from.
     *
     * A pasted `smb://host/share/path` or `\\host\share\path` in the host field is read
     * whole, because that is the form a reader is most likely to have to hand.
     */
    fun connect() {
        val pasted = SmbAddress.parse(host.value)
        val target = when {
            pasted != null && share.value.isBlank() -> pasted.copy(
                username = username.value.ifBlank { null },
                password = password.value.ifBlank { null },
            )
            else -> {
                // A host may carry a port. Most readers never type one, but a NAS behind a
                // forwarded port has no other way to say so.
                val typed = host.value.trim().removePrefix("smb://").trim('/', '\\')
                val parts = typed.split(':')
                SmbAddress(
                    host = parts[0],
                    share = share.value.trim().trim('/', '\\'),
                    username = username.value.ifBlank { null },
                    password = password.value.ifBlank { null },
                    port = parts.getOrNull(1)?.toIntOrNull() ?: SmbAddress.DEFAULT_PORT,
                )
            }
        }
        if (target.host.isBlank() || target.share.isBlank()) {
            _step.value = Step.Failed(context.getString(R.string.smb_error_not_an_address))
            return
        }

        _step.value = Step.Connecting
        // On IO from the first line: jcifs resolves names while it builds a client, and
        // doing that on the main thread is an ANR rather than a slow screen.
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val client = SmbClient(target)
                val identity = client.connect()
                resolved = target
                _step.value = Step.Browsing(identity, "", client.list(""))
            } catch (error: SmbError) {
                _step.value = Step.Failed(describe(error))
            }
        }
    }

    /** Opens a folder inside the share, so the reader can go on choosing. */
    fun enter(path: String) {
        val target = resolved ?: return
        val current = _step.value as? Step.Browsing ?: return
        viewModelScope.launch(Dispatchers.IO) {
            val entries = runCatching { SmbClient(target).list(path) }.getOrNull()
            _step.value = entries?.let { current.copy(path = path, entries = it) } ?: current
        }
    }

    /** The folder above the one being shown, or null at the share's root. */
    fun parentOf(path: String): String? =
        path.takeIf { it.isNotEmpty() }?.substringBeforeLast('/', "")

    /**
     * The source to save, rooted at the folder the reader chose.
     *
     * Null when the password cannot be stored, and the step says so. A share whose password
     * is gone is a row that fails on the next launch with nothing to explain why.
     */
    fun source(): Source? {
        val target = resolved ?: return null
        val current = _step.value as? Step.Browsing ?: return null
        val rooted = target.copy(path = current.path)

        val id = UUID.randomUUID()
        var reference: String? = null
        if (!rooted.isGuest) {
            val key = CredentialStore.reference(id)
            if (credentials == null || !credentials.save(rooted.password.orEmpty(), key)) {
                _step.value = Step.Failed(context.getString(R.string.smb_error_key_not_stored))
                return null
            }
            reference = key
        }

        return Source(
            id = id,
            displayName = rooted.displayName,
            kind = SourceKind.NETWORK_SHARE,
            state = SourceConnectionState.Connected,
            credentialReference = reference,
            locator = SmbLocator.of(rooted),
        )
    }

    fun reset() {
        _step.value = Step.Entering
    }

    private fun describe(error: SmbError): String = context.getString(
        when (error) {
            is SmbError.HostUnreachable -> R.string.smb_error_host_unreachable
            is SmbError.ShareNotFound -> R.string.smb_error_share_not_found
            is SmbError.AuthenticationRejected -> R.string.smb_error_authentication
            is SmbError.ProtocolUnsupported -> R.string.smb_error_smb1
            is SmbError.Unexpected -> R.string.smb_error_unexpected
        },
    )
}

/**
 * A share written down, and read back.
 *
 * The registry stores one string per source, so the host, share, path and user name travel
 * as a URL. The password never does -- that is what the credential store is for.
 */
object SmbLocator {
    fun of(address: SmbAddress): String {
        val user = address.username?.takeIf { it.isNotEmpty() }?.let { "$it@" } ?: ""
        val port = if (address.port == SmbAddress.DEFAULT_PORT) "" else ":${address.port}"
        val path = address.path.trim('/').let { if (it.isEmpty()) "" else "/$it" }
        return "smb://$user${address.host}$port/${address.share}$path"
    }

    fun parse(locator: String, password: String?): SmbAddress? {
        val body = locator.removePrefix("smb://")
        val user = body.substringBefore('@', "").takeIf { body.contains('@') }
        val rest = if (user != null) body.substringAfter('@') else body
        val parsed = SmbAddress.parse(rest) ?: return null
        return parsed.copy(username = user, password = password)
    }
}

/** What is needed to open a saved share. */
data class SmbPage(val id: String, val title: String, val address: SmbAddress) {
    companion object {
        /** Null when the source is not a share, has no address, or has lost its password. */
        fun of(source: Source, credentials: CredentialStore?): SmbPage? {
            if (source.kind != SourceKind.NETWORK_SHARE) return null
            val locator = source.locator ?: return null
            val password = source.credentialReference?.let { credentials?.secret(it) }
            if (source.credentialReference != null && password == null) return null
            val address = SmbLocator.parse(locator, password) ?: return null
            return SmbPage(source.id.toString(), source.displayName, address)
        }
    }
}

File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/SmbConnection.swift
public import Foundation

public import Persistence
public import Smb
public import StoryArcCore

/// Adding a share, from a host and a password to a saved source.
///
/// `network-share` validates "before saving" and reports "the specific failure", then lets a
/// reader "browse the share's directory tree and pick the folder to use as the library
/// root". Those are two steps rather than one screen, and this is the state machine between
/// them. Android's `SmbConnection` is the same one.
@Observable
@MainActor
public final class SmbConnection {
    public enum Step: Equatable, Sendable {
        case entering
        case connecting
        /// Connected. The reader is now choosing which folder to read.
        case browsing(SmbIdentity, path: String, entries: [SmbEntry])
        case failed(String)
    }

    public private(set) var step: Step = .entering

    public var host = ""
    public var share = ""
    public var username = ""
    public var password = ""

    private let credentials: CredentialStore?
    private var resolved: SmbAddress?

    public init(credentials: CredentialStore? = nil) {
        self.credentials = credentials
    }

    /// Whether there is enough typed in to try.
    public var canConnect: Bool {
        guard step != .connecting else { return false }
        let typedHost = host.trimmingCharacters(in: .whitespaces)
        guard !typedHost.isEmpty else { return false }
        // A pasted `\\host\share` names both at once, so the share field may be empty.
        return !share.trimmingCharacters(in: .whitespaces).isEmpty
            || SmbAddress.parse(typedHost) != nil
    }

    /// Connects and lists the share's root, which is the first thing a reader chooses from.
    public func connect() async {
        guard let target = typedAddress() else {
            step = .failed(String(localized: "smb.error.notAnAddress", bundle: .module, locale: .storyArc))
            return
        }

        step = .connecting
        do {
            let client = SmbClient(address: target)
            let identity = try await client.connect()
            resolved = target
            step = .browsing(identity, path: target.path, entries: try await client.list(target.path))
        } catch let error as SmbError {
            step = .failed(Self.describe(error))
        } catch {
            step = .failed(String(localized: "smb.error.unexpected", bundle: .module, locale: .storyArc))
        }
    }

    /// Opens a folder inside the share, so the reader can go on choosing.
    public func enter(_ path: String) async {
        guard let target = resolved, case let .browsing(identity, _, previous) = step else { return }
        let entries = (try? await SmbClient(address: target).list(path)) ?? previous
        step = .browsing(identity, path: path, entries: entries)
    }

    /// The folder above the one being shown, or nil at the share's root.
    public func parent(of path: String) -> String? {
        guard !path.isEmpty else { return nil }
        guard let cut = path.lastIndex(of: "/") else { return "" }
        return String(path[path.startIndex..<cut])
    }

    /// The source to save, rooted at the folder the reader chose.
    ///
    /// Nil when the password cannot be stored, and the step says so. A share whose password
    /// is gone is a row that fails on the next launch with nothing to explain why.
    public func source() -> Source? {
        guard let target = resolved, case let .browsing(_, path, _) = step else { return nil }
        let rooted = SmbAddress(
            host: target.host,
            share: target.share,
            path: path,
            username: target.username,
            password: target.password,
            port: target.port
        )

        let id = UUID()
        var reference: String?
        if !rooted.isGuest {
            let key = CredentialStore.reference(for: id)
            guard credentials?.save(rooted.password ?? "", for: key) == true else {
                step = .failed(String(localized: "smb.error.keyNotStored", bundle: .module, locale: .storyArc))
                return nil
            }
            reference = key
        }

        return Source(
            id: id,
            displayName: rooted.displayName,
            kind: .networkShare,
            state: .connected,
            credentialReference: reference,
            locator: SmbLocator.write(rooted)
        )
    }

    public func reset() {
        step = .entering
    }

    private func typedAddress() -> SmbAddress? {
        let typedHost = host.trimmingCharacters(in: .whitespaces)
        let typedShare = share.trimmingCharacters(in: .whitespaces)
        let user = username.isEmpty ? nil : username
        let secret = password.isEmpty ? nil : password

        if typedShare.isEmpty, let pasted = SmbAddress.parse(typedHost) {
            return SmbAddress(
                host: pasted.host,
                share: pasted.share,
                path: pasted.path,
                username: user,
                password: secret,
                port: pasted.port
            )
        }

        // A host may carry a port. Most readers never type one, but a NAS behind a
        // forwarded port has no other way to say so.
        let parts = typedHost
            .replacingOccurrences(of: "smb://", with: "")
            .trimmingCharacters(in: CharacterSet(charactersIn: "/\\"))
            .split(separator: ":")
            .map(String.init)
        guard let bare = parts.first, !bare.isEmpty, !typedShare.isEmpty else { return nil }

        return SmbAddress(
            host: bare,
            share: typedShare,
            username: user,
            password: secret,
            port: parts.count > 1 ? Int(parts[1]) ?? SmbAddress.defaultPort : SmbAddress.defaultPort
        )
    }

    private static func describe(_ error: SmbError) -> String {
        switch error {
        case .hostUnreachable:
            String(localized: "smb.error.hostUnreachable", bundle: .module, locale: .storyArc)
        case .shareNotFound:
            String(localized: "smb.error.shareNotFound", bundle: .module, locale: .storyArc)
        case .authenticationRejected:
            String(localized: "smb.error.authentication", bundle: .module, locale: .storyArc)
        case .protocolUnsupported:
            String(localized: "smb.error.smb1", bundle: .module, locale: .storyArc)
        case .encryptionRequired:
            String(localized: "smb.error.encryption", bundle: .module, locale: .storyArc)
        case .unexpected:
            String(localized: "smb.error.unexpected", bundle: .module, locale: .storyArc)
        }
    }
}

/// A share written down, and read back.
///
/// The registry stores one string per source, so the host, share, path and user name travel
/// as a URL. The password never does — that is what the credential store is for.
public enum SmbLocator {
    public static func write(_ address: SmbAddress) -> String {
        let user = (address.username?.isEmpty == false) ? "\(address.username ?? "")@" : ""
        let port = address.port == SmbAddress.defaultPort ? "" : ":\(address.port)"
        let trimmed = address.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let path = trimmed.isEmpty ? "" : "/\(trimmed)"
        return "smb://\(user)\(address.host)\(port)/\(address.share)\(path)"
    }

    public static func read(_ locator: String, password: String?) -> SmbAddress? {
        let body = locator.replacingOccurrences(of: "smb://", with: "")
        let user = body.contains("@") ? String(body.split(separator: "@")[0]) : nil
        let rest = user != nil ? String(body.split(separator: "@").dropFirst().joined(separator: "@")) : body
        guard let parsed = SmbAddress.parse(rest) else { return nil }
        return SmbAddress(
            host: parsed.host,
            share: parsed.share,
            path: parsed.path,
            username: user,
            password: password,
            port: parsed.port
        )
    }
}

/// What is needed to open a saved share.
public struct SmbPage: Sendable {
    public let id: String
    public let title: String
    public let address: SmbAddress

    /// Nil when the source is not a share, has no address, or has lost its password.
    public init?(source: Source, credentials: CredentialStore?) {
        guard source.kind == .networkShare, let locator = source.locator else { return nil }
        let password = source.credentialReference.flatMap { credentials?.secret(for: $0) }
        if source.credentialReference != nil, password == nil { return nil }
        guard let address = SmbLocator.read(locator, password: password) else { return nil }

        id = source.id.uuidString
        title = source.displayName
        self.address = address
    }
}
